
<div style="border:solid green 2px; padding: 20px">

👋 **¡Hola! Soy Dot, tu revisor de IA.**

He completado la primera revisión de tu código. A continuación, encontrarás mis comentarios y sugerencias de mejora.

**¿Qué debes hacer ahora?**

1. **Revisar:** Lee mis comentarios en el notebook más abajo.
2. **Decidir:**
* **¿Estás de acuerdo con el feedback?** ¡Genial! Actualiza tu código según las sugerencias.
* **¿No estás de acuerdo o crees que es un error?** ¡No hay problema! Puedes dejar tu código exactamente como está.


3. **Volver a enviar:** Envía tu proyecto de nuevo. **Tu próxima iteración será revisada por un experto humano**, independientemente de si hiciste cambios o no.

-------------------------------------------------------------------------------------------------------------
Mis comentarios están más abajo. **Te pido amablemente que no los muevas, modifiques ni elimines**.

Verás mis comentarios resaltados en cuadros verdes, amarillos o rojos como estos:

<div class="alert alert-block alert-success">
<b>Reviewer's comment, Iteration X</b> <a class="tocSkip"></a>

Correcto. Todo se ha hecho exitosamente.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment, Iteration X</b> <a class="tocSkip"></a>

Observaciones. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">
<b>Reviewer's comment, Iteration X</b> <a class="tocSkip"></a>

Necesita corrección. El bloque requiere algunas correcciones.
</div>



# Déjame escuchar música

Los datos están almacenados en el archivo `/datasets/music_project_en.csv`.

¡Ahora sí, manos al código!


## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Etapa 1.1. Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importa pandas
import pandas as pd

<details>
<summary>Haz clic para ver la pista</summary>

Para usar `pandas`, necesitas la palabra clave `import` seguida del nombre de la biblioteca y un alias.


Etapa 1.2. Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
# Lee el archivo y almacénalo en df
df = pd.read_csv('/datasets/music_project_en.csv')

<div class="alert alert-block alert-success">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Muy buena configuración del proyecto: importaste pandas correctamente y cargaste el dataset desde la ruta esperada en `df`.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa la palabra clave `read_csv` para leer el archivo CSV. Asegúrate de incluir la ruta `/datasets/music_project_en.csv` y guarda el resultado en la variable `df`.

Etapa 1.3. Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtén las 10 primeras filas de la tabla df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `head` para ver las primeras filas del DataFrame.

Etapa 1.4. Obtén la información general sobre la tabla con el método info().

In [4]:
# Obtén la información general sobre nuestros datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `info` para obtener detalles sobre columnas, tipos de datos y valores nulos.


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la canción;
- `'City'`: ciudad del usuario;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver dos problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.

2. `Identifica tú mismo el segundo problema y escríbelo aquí.`

   Hay nombres de las columnas que empiezan o terminan por mayúscula, otros contienen espacios vacios. Habría que normalizar la representación de las columnas para que sea más sencillo el tratamiento y las llamadas a los datos en el análisis. Los problemas están en ' userID','Track','City','Day'

<div class="alert alert-block alert-warning">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

En tus observaciones sobre la tabla, intenta referirte a los nombres de las columnas exactamente como aparecen en la salida (incluyendo cualquier espacio al inicio/final y la capitalización). Esto importa porque esos problemas exactos en los encabezados son los que corriges en los pasos de preprocesamiento.

</div>


### Escribe algunas observaciones por tu parte. Contesta a las siguientes preguntas: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

* Hay solamente datos tipo object, son strings en celdas del dataframe. Cuando se llama a dtypes devuelve en la columna Dtype el tipo de dato.

<div class="alert alert-block alert-warning">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Esta sección te pide que respondas tres preguntas, pero solo abordaste la primera. Considera añadir notas breves sobre (2) si el dataset parece suficiente para probar la hipótesis y (3) qué problemas notaste (p. ej., valores faltantes, duplicados o tipos de datos que puedan necesitar conversión).

</div>


* Analizo las observaciones propuestas argumentando con código,outpout y mark-down más abajo.                                                                                        

In [5]:
# Compruebo los tipos de datos
df.dtypes

  userID    object
Track       object
artist      object
genre       object
  City      object
time        object
Day         object
dtype: object

* Visualizo datos estadisticos sobre el df para tener una impresión general

In [6]:
df.describe(include = 'object')

,userID,Track,artist,genre,City,time,Day
count,65079,63736,57512,63881,65079,65079,65079
unique,41748,39666,37806,268,2,20392,3
top,A8AE9169,Brand,Kartvelli,pop,Springfield,08:14:07,Friday
freq,76,136,136,8850,45360,14,23149


* Compruebo el número de duplicados explicitos, quiere decir que coinciden en todas las columnas de la fila con otra muestra, hay que tratar los datos más adelante para un análisis preciso.

In [7]:

# Compruebo la cantidad de valores duplicados
# Indica la cantidad de duplicados explicitos (todos los valores de las columnas de la fila idénticos)
print(f'Numero de duplicados explicitos : {df.duplicated().sum()}')


Numero de duplicados explicitos : 3826


* Visualizo la cantidad de datos únicos para una impresión general
* Relevante para el análisis : el df cuenta con muestras de 2 ciudades, 3 dias de la semana y 268 géneros musicales

In [8]:
# Cantidad de datos únicos por columna
print(df.nunique())

  userID    41748
Track       39666
artist      37806
genre         268
  City          2
time        20392
Day             3
dtype: int64


* Más adelante si fuera necesario para el analisis, recomendaría cambiar el tipo de dato para la columna de "time", es el unico dato que sería útil que fuera tipo número, int o float según la conversión o aproximación que se necesite.

In [9]:
# hay que indagar para buscar resultados duplicados implícitos.
# df['genre'].sort_values().unique()
# He revisado por encima la columna genre (me resulta un dato relevante)
# En un primer vistazo no he visto duplicados implicitos.
# No muestro la salida para no saturar el cuaderno con una lista gigante

# Datos que son object pero podrían tratarse como float para un tratamiento más eficaz.
# Se podría "traducir" time a un numero entero o una aproximación float para analizar que género es más escuchado según el momento del día
# El tipo de dato en object-str resultaría complejo de utilizar para el análisis
df['time']

0        20:28:33
1        14:07:09
2        20:58:07
3        08:37:09
4        08:34:34
           ...   
65074    13:32:28
65075    10:00:00
65076    20:09:26
65077    21:43:59
65078    21:59:46
Name: time, Length: 65079, dtype: object

## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>

Tu objetivo aquí es preparar los datos para analizarlos.
El primer paso es resolver los problemas con los encabezados. Después podemos avanzar a los valores ausentes y duplicados. ¡Empecemos!

Vamos a corregir el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Etapa 2.1. Muestra los encabezados de la tabla (los nombres de las columnas):

In [10]:
# Muestra los nombres de las columnas
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

<details>
<summary>Haz clic para ver la pista</summary>

Usa el atributo `.columns` para ver todos los nombres de las columnas en el DataFrame.


Vamos cambiar los encabezados de la tabla siguiendo las reglas estilísticas convencionales:
*   Todos los caracteres deben ser minúsculas.
*   Elimina los espacios.
*   Si el nombre tiene varias palabras, utiliza snake_case, es decir, añade un guion bajo ( _ ) entre las palabras en lugar de un espacio.



Etapa 2.2. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [11]:
# Bucle que itera sobre los encabezados y los pone todos en minúsculas
nuevas_columnas=[]
for column in df.columns:
    nuevas_columnas.append(column.lower())
df.columns = nuevas_columnas

df.columns

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')

<details>
<summary>Haz clic para ver la pista</summary>

Usa un `for` para recorrer `df.columns` y aplica `.lower()` a cada nombre. Luego, asigna la lista resultante de nuevo a `df.columns`.


Etapa 2.3. Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas y muestra los nombres de las columnas de nuevo:

In [12]:
# Bucle que itera sobre los encabezados y elimina los espacios
nuevas_columnas=[]
for column in df.columns:
    nuevas_columnas.append(column.strip())
df.columns = nuevas_columnas

df.columns

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

<details>
<summary>Haz clic para ver la pista</summary>

Usa `.strip()` dentro del bucle para limpiar los nombres de columnas. Asegúrate de reasignar los nuevos nombres a `df.columns`.


Etapa 2.4. Necesitamos aplicar la regla de snake_case en la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [13]:
# Cambia el nombre de la columna "userid"
nuevas_columnas = {'userid':'user_id'}
df.rename(columns = nuevas_columnas, inplace = True)
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

Chequeo el resultado y el nombre de las columnas ya aparece como se requiere.

<div class="alert alert-block alert-warning">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Este paso ya está completado: renombraste `userid` a `user_id` y mostraste `df.columns`, lo cual confirma el cambio. El paso adicional de “check the result” justo después servirá como la verificación formal.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `.rename()` con el parámetro `columns` para cambiar `userid` a `user_id`. No olvides establecer `inplace=True` o reasignar a `df`.

Etapa 2.5. Comprueba el resultado. Muestra los encabezados una vez más:

In [14]:
# Comprueba el resultado: lista de encabezados
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.columns` para volver a ver los nombres actualizados de las columnas.

### Valores ausentes <a id='missing_values'></a>
 Etapa 2.5. Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos para obtener el número de valores ausentes.

In [15]:
# Calcula el número de valores ausentes
# df.isnull().sum()
df.isna().sum()

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

<div class="alert alert-block alert-danger">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Este paso pide dos métodos para medir los valores faltantes, pero solo usaste uno (`isnull().sum()`). Considera añadir una segunda verificación diferente (por ejemplo, usando `isna()` como alternativa, o calculando el número total de valores faltantes en todo el DataFrame) para cumplir completamente con el requisito.

</div>


<div class="alert alert-block alert-success">
<b>Reviewer's comment, Iteration 2</b> <a class="tocSkip"></a>

Muy bien hecho Javier, ambos métodos son adecuados para el caso, felicitaciones!

<details>
<summary>Haz clic para ver la pista</summary>

Usa `.isna()` para detectar los valores ausentes y combínalo con el método `.sum()` para contar cuántos hay por columna.


Etapa 2.6. Sustituye los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`.

1. Crea una lista llamada columns_to_replace que contenga los nombres de las columnas 'track', 'artist' y 'genre'.

2. Usa un bucle for para iterar sobre cada columna en columns_to_replace.

3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.

In [16]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
columns_to_replace = ['track','artist','genre']
for column in columns_to_replace:
    df[column].fillna('unknown', inplace = True)

<details>
<summary>Haz clic para ver la pista</summary>

Crea una lista con los nombres de las columnas y usa un `for` para recorrerla. Dentro del bucle, aplica `.fillna('unknown')` a cada una.

Etapa 2.7. Ahora comprueba el resultado para asegurarte de que no falten valores ausentes por reemplazar en el conjunto de datos. Para ello, cuenta los valores ausentes una vez más.

In [17]:
# Cuenta los valores ausentes
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

<div class="alert alert-block alert-success">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Buen trabajo al validar tu corrección: reemplazaste los valores faltantes en las columnas solicitadas y luego volviste a revisar los conteos de null para confirmar que todo está completo.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa `isna()` con `sum()` otra vez para confirmar que ya no quedan valores nulos en las columnas modificadas.

### Duplicados <a id='duplicates'></a>
Etapa 2.8. Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos para obtener la cantidad de duplicados explícitos.

In [21]:
# Compruebo el tamaño del dataframe
print(f'Tamaño del df antes de hacer cambios (filas,columnas) : {df.shape}')

Tamaño del df antes de hacer cambios (filas,columnas) : (65079, 7)


In [ ]:
# Cuenta los duplicados explícitos
df.duplicated().sum()

<div class="alert alert-block alert-danger">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Este paso también requiere dos métodos para contar duplicados explícitos, pero solo se muestra uno (`df.duplicated().sum()`). Añade un segundo enfoque (p. ej., comparar el tamaño del DataFrame antes/después de eliminar duplicados) para cumplir la instrucción.

</div>


<div class="alert alert-block alert-warning">
<b>Reviewer's comment, Iteration 2</b> <a class="tocSkip"></a>

Aquí se nos plantea poder mostrar dos enfoques para chequear duplicados, sin embargo no es suficiente para marcarlo como error. El metodo duplicated() es más que acorde seguido del sum()

<details>
<summary>Haz clic para ver la pista</summary>

Usa un método que detecte filas repetidas (`duplicated`) y luego otro para contarlas.

Etapa 2.9. Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [22]:
# Elimina los duplicados explícitos
df.drop_duplicates(inplace=True)

<details>
<summary>Haz clic para ver la pista</summary>

Busca un método que elimine las filas duplicadas directamente del DataFrame. No olvides actualizar la variable.

Etapa 2.10. Comprobemos ahora si conseguimos eliminar todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [23]:
# Comprueba de nuevo si hay duplicados
df.duplicated().sum()

0

In [24]:
# Compruebo el tamaño del dataframe después del cambio
print(f'Tamaño del df después de hacer cambios (filas,columnas) : {df.shape}')

Tamaño del df después de hacer cambios (filas,columnas) : (61253, 7)


Efectivamente el numero de muestras ha disminuido pero eso convierte los futuros resultados del análisis en más precisos

<details>
<summary>Haz clic para ver la pista</summary>

Repite el mismo enfoque que usaste para contar duplicados antes. Deberías ver que el total ahora es cero.

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Etapa 2.11. Primero debemos mostrar una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.


In [25]:
# Inspecciona los nombres de géneros únicos
# Ordeno con sort_values()
df['genre'].sort_values().unique()

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'ambient', 'americana', 'animated', 'anime',
       'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook',
       'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom',
       'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian',
       'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop',
       'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber',
       'children', 'chill', 'chinese', 'choral', 'christian', 'christmas',
       'classical', 'classicmetal', 'club', 'colombian', 'comedy',
       'conjazz', 'contemporary', 'country', 'cuban', 'dance',
       'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock',
       'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat',
       'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy',
       'electronic', 'electropop', 'emo', 'entehno', 'epicmetal',
       'estrada', 'ethnic', 'eurofo

<details>
<summary>Haz clic para ver la pista</summary>

Usa `unique()` para obtener los géneros únicos. Si quieres verlos en orden alfabético, combina con `sorted()`.

Etapa 2.12. Vamos a examinar la lista para identificar **duplicados implícitos** del género `hiphop`, es decir, nombres mal escritos o variantes que hacen referencia al mismo género musical.

Los duplicados que encontrarás son:

* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo, vamos a crear una función llamada `replace_wrong_values()`.


1. Define una función llamada `replace_wrong_values()` que reciba los siguientes parámetros:

* `df`: el DataFrame a modificar
* `column`: el nombre de la columna a trabajar
* `wrong_values`: una lista con los valores incorrectos
* `correct_value`: el valor correcto para reemplazar

2. Dentro de la función, usa un bucle `for` para iterar sobre cada valor incorrecto y aplicar `.replace()`.


In [26]:
# Función para reemplazar los duplicados implícitos
def replace_wrong_values(df,column,wrong_values,correct_value):
    for wrong_value in wrong_values:
        df[column] = df[column].replace(wrong_value,correct_value)

<div class="alert alert-block alert-danger">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

La función funciona, pero la consigna te pide específicamente que uses un bucle `for` dentro de `replace_wrong_values()` para iterar sobre `wrong_values` y aplicar `.replace()` por cada valor. Considera reescribir la función para que se ajuste a ese enfoque solicitado.

</div>


<div class="alert alert-block alert-success">
<b>Reviewer's comment, Iteration 2</b> <a class="tocSkip"></a>

Excleente estructura de la función Javier, tomando los parametros correspondientes e iterando mediante un for los valores que queremos corregir luego con replace(), incríble!

<details>
<summary>Haz clic para ver la pista</summary>

Crea una función con cuatro parámetros. Dentro del bucle, llama a `.replace()` sobre la columna para cada valor incorrecto, cambiándolo por el correcto.

Etapa 2.13. Ahora, llama a la función pasando:

* `df` como el DataFrame
* `'genre'` como nombre de columna
* `['hip', 'hop', 'hip-hop']` como lista de valores incorrectos
* `'hiphop'` como valor correcto


In [27]:
# Elimina los duplicados implícitos
column = 'genre'
wrong_values = ['hip', 'hop', 'hip-hop']
correct_value = 'hiphop'
replace_wrong_values(df, column, wrong_values, correct_value)

<details>
<summary>Haz clic para ver la pista</summary>

Llama a la función que creaste y pasa los cuatro argumentos necesarios. Asegúrate de usar el nombre correcto de la columna y lista los valores que deben unificarse.

Etapa 2.14. Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [28]:
# Comprueba de nuevo los duplicados implícitos
df['genre'].sort_values().unique()

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'ambient', 'americana', 'animated', 'anime',
       'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook',
       'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom',
       'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian',
       'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop',
       'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber',
       'children', 'chill', 'chinese', 'choral', 'christian', 'christmas',
       'classical', 'classicmetal', 'club', 'colombian', 'comedy',
       'conjazz', 'contemporary', 'country', 'cuban', 'dance',
       'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock',
       'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat',
       'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy',
       'electronic', 'electropop', 'emo', 'entehno', 'epicmetal',
       'estrada', 'ethnic', 'eurofo

<details>
<summary>Haz clic para ver la pista</summary>

Usa de nuevo el método que muestra los valores únicos de una columna para verificar que todos los nombres se hayan unificado correctamente.

## Etapa 3. Análisis

### Tarea: Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

Queremos analizar si hay diferencias en la cantidad de canciones reproducidas en Springfield y Shelbyville. Para ello, usaremos los datos de dos días de la semana: lunes y viernes.

Compararemos cuántas canciones se escucharon en cada ciudad durante esos días para identificar posibles patrones de comportamiento.

Sigue estos tres pasos para organizar tu análisis:

- Dividir: agrupa los datos por ciudad.

- Aplicar: cuenta cuántas canciones se reproducen en cada grupo.

- Combinar: presenta los resultados de forma que se puedan comparar fácilmente ambas ciudades.

Repite este proceso por separado para cada uno de los dos días.

Etapa 3.1.
Cuenta cuántas canciones se reprodujeron en cada ciudad utilizando la columna `'track'` como referencia.

In [29]:
# Cuenta las canciones reproducidas en cada ciudad
filtro_ciudades = df[df['city'].isin(['Springfield','Shelbyville'])]
filtro_ciudades.groupby('city')['track'].count()

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64

Hay más escuchas en Springfield que en Shelbyville, seguramente Shelbyville tenga menos habitantes.

<details>
<summary>Haz clic para ver la pista</summary>

Agrupa los datos por ciudad con el método `groupby()` y usa el método `count()` para contar las canciones reproducidas en cada grupo.

Etapa 3.2. Redacta brevemente tus observaciones sobre los resultados.

¿Qué diferencias encontraste entre Springfield y Shelbyville? ¿A qué podrían deberse?

Etapa 3.3.
Agrupa los datos por día de la semana y cuenta cuántas canciones se reprodujeron los lunes y viernes.



In [30]:
# Calcula las canciones reproducidas en cada uno de los dos días
# Observo como están nombrados los días de la semana
print(df['day'].unique())
print()
# Divido el df para días lunes y viernes
filtro_dias = df[df['day'].isin(['Monday','Friday'])]
# Combino los datos de dia y track y los cuento
filtro_dias.groupby('day')['track'].count()

['Wednesday' 'Friday' 'Monday']



day
Friday    21840
Monday    21354
Name: track, dtype: int64

La diferencia entre el número de escuchas del lunes comparadas con el viernes es mínima, el día de la semana no parece ser relevante para el número de escuchas.

<div class="alert alert-block alert-warning">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Esta parte pide una breve interpretación escrita comparando el lunes vs el viernes. Calculaste los conteos, pero considera agregar una observación corta en markdown que explique lo que sugieren los números.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Cambia el criterio de agrupación a la columna que representa el día de la semana y filtra solo para los días que te interesan.

Etapa 3.4. Describe brevemente qué observaste al comparar los lunes y viernes.

¿Hubo un día con más actividad? ¿Cambia algo si analizas cada ciudad por separado?

In [31]:
# Solicito los días y ciudades requeridas
filtro_alternativo = df.query("day in ['Monday','Friday'] and city in ['Shelbyville','Springfield']")
# Combino los dias y ciudades solicitados con track, y cuento los casos
filtro_alternativo.groupby(['day','city'])['track'].count()

day     city       
Friday  Shelbyville     5895
        Springfield    15945
Monday  Shelbyville     5614
        Springfield    15740
Name: track, dtype: int64

Como observamos anteriormente en Shelbyville hay menos reproducciones de canciones, seguramente tenga menos población, también no hay diferecias significativas en el conteo según el día que se observe la muestra.

<div class="alert alert-block alert-warning">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Problema recurrente: cuando un paso pide una conclusión escrita, conserva el código de comparación, pero asegúrate de que la interpretación esté claramente expresada en la celda de respuesta en markdown correspondiente para que sea fácil verificar que la tarea está completa.

</div>


Etapa 3.5

Ahora vamos a combinar dos criterios: día y ciudad.

Crea una función llamada `number_tracks()` que reciba dos parámetros:

* `day`: un día de la semana (por ejemplo, `'Monday'`)
* `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`)
  
Dentro de la función:

1. Filtra el DataFrame por el día.
2. Luego, filtra por la ciudad.
3. Cuenta cuántas veces aparece `'user_id'` en ese filtro.
4. Devuelve ese número como resultado.


In [32]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(day,city):
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    filtro_a = df.loc[df['day'] == day]
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    filtro_requerido = filtro_a.loc[filtro_a['city'] == city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    resultado = filtro_requerido['user_id'].count()
    # Devuelve el número de valores de la columna 'user_id'
    return resultado

<details>
<summary>Haz clic para ver la pista</summary>

Define una función con dos parámetros. Dentro, aplica un filtrado secuencial usando `df[...]`, y al final cuenta los valores con `.count()`.

Etapa 3.6. Llama a `number_tracks()` cuatro veces: una por ciudad en cada uno de los dos días.

In [33]:
# El número de canciones reproducidas en Springfield el lunes
print(number_tracks('Monday','Springfield'))

15740


In [34]:
# El número de canciones reproducidas en Shelbyville el lunes
print(number_tracks('Monday','Shelbyville'))

5614


In [35]:
# El número de canciones reproducidas en Springfield el viernes
print(number_tracks('Friday','Springfield'))

15945


In [36]:
# El número de canciones reproducidas en Shelbyville el viernes
print(number_tracks('Friday','Shelbyville'))

5895


<div class="alert alert-block alert-success">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Bien hecho: `number_tracks()` sigue la lógica requerida de filtrado y conteo, y tus cuatro llamadas producen valores claros que respaldan la comparación por ciudad/día.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Llama a tu función con combinaciones distintas de ciudad y día, como `'Monday', 'Springfield'`, y guarda o imprime los resultados para comparar.

# Número de escuchas según género

In [37]:
# Análisis preliminar
# Agrupo según genero para ver los casos
filtro_genero = df.groupby('genre')['track'].count()

filtro_genero = filtro_genero.sort_values(ascending = False)
# Muestra los 3 generos más escuchados
print(filtro_genero.head(3))
print()
# Muestra los 3 menos escuchados
print(filtro_genero.tail(3))


genre
pop      8323
dance    6367
rock     5844
Name: track, dtype: int64

genre
kayokyoku    1
karadeniz    1
ïîï          1
Name: track, dtype: int64


# Conclusiones <a id='end'></a>

## Escribe tus conclusiones finales sobre el análisis

Redacta un resumen breve y claro de los hallazgos obtenidos durante el proceso de análisis. Tu conclusión debe:

* Mencionar los principales patrones que observaste en los datos.
* Identificar cualquier problema encontrado y cómo lo solucionaste.
* Explicar cómo estas acciones ayudaron a mejorar la calidad del análisis.

Reflexiona también sobre la pregunta central:

> *¿Los datos muestran que el comportamiento de los usuarios —en cuanto a la música que escuchan— varía según la ciudad y el día de la semana?*

Apoya tu respuesta con ejemplos concretos de los resultados obtenidos.


# Los datos obtenidos del analisis indican que la cantidad de escuchas es indiferente al día de la semana, sin embargo en cuanto a las ciudades se observa que la cantidad de escuchas es significativamente diferente (la principal hipótesis es la diferencia de población pero sin los datos es dificil saberlo)

* Selbyville tiene menos reproducciones que Springfield, independientemente del día de la semana el número de escuchas es similar.
  
* He normalizado los nombres de las columnas para evitar errores sintácticos al hacer consultas, he sustituido valores ausentes en columnas para facilitar el analisis.

* He eliminado valores duplicados explicitos(valores iguales en todas los valores de la fila).

* He detectado y corregido valores duplicados implícitos en una columna del dataframe. Esto ayuda a tener orden y cuando se hagán consultas sobre un valor determinado sobre el concepto de la columna sea preciso y no se pierdan muestras.

* El tratamiento de valores duplicados en el caso de este dataframe ayuda a tener más precisión a la hora de hacer conteos de datos, dependiendo de la columna que apuntemos al hacer la consulta si hubiera duplicados sería impreciso.

In [38]:
# Cuenta las canciones reproducidas en cada ciudad
filtro_ciudades = df[df['city'].isin(['Springfield','Shelbyville'])]
filtro_ciudades.groupby('city')['track'].count()

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64

* Sobre la cantidad de escuchas por ciudad, se observa que hay más reproducciones de canciones en Springfield

In [39]:
# Solicito los días y ciudades requeridas
filtro_alternativo = df.query("day in ['Monday','Friday'] and city in ['Shelbyville','Springfield']")
# Combino los dias y ciudades solicitados con track, y cuento los casos
filtro_alternativo.groupby(['day','city'])['track'].count()

day     city       
Friday  Shelbyville     5895
        Springfield    15945
Monday  Shelbyville     5614
        Springfield    15740
Name: track, dtype: int64

* Independientemente del día de la semana el numero de escuchas es similar en ambas ciudades

<div class="alert alert-block alert-warning">
<b>Reviewer's comment, Iteration 1</b> <a class="tocSkip"></a>

Tus conclusiones son un buen punto de partida, pero serían más sólidas si respondieras explícitamente a la pregunta central (¿varía el comportamiento por ciudad y por día?) y lo respaldaras con cifras concretas de tus resultados (p. ej., los conteos de lunes/viernes por ciudad). Considera también resumir brevemente las principales correcciones de calidad de datos que realizaste (headers, valores faltantes, duplicados) y cómo cada una mejora la confiabilidad del análisis.

</div>
